### PR TITLE
Add hotpluging testing with discard disk

### PR DESCRIPTION
**Pull request by qingwangrh**
ID:1892582,1892581
Signed-off-by: qingwangrh <qinwang@redhat.com>

### CODE REVIEW

**Comment by qingwangrh** [comment]
python ConfigTest.py --testcase=block_discard_hotplug --guestname=RHEL.8.3.0,Win2019 --driveformat=virtio_blk,virtio_scsi --machine=i440fx,q35 --imageformat=raw,qcow2 --clone=no
(01/12) Host_RHEL.m8.u2.product_av.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.block_discard_hotplug.local_file: PASS (132.40 s)
 (02/12) Host_RHEL.m8.u2.product_av.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.block_discard_hotplug.local_file.q35: PASS (134.32 s)
 (03/12) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.block_discard_hotplug.local_file: PASS (129.45 s)
 (04/12) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.block_discard_hotplug.local_file.q35: PASS (134.14 s)
 (05/12) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.block_discard_hotplug.scsi_debug: PASS (125.88 s)
 (06/12) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.block_discard_hotplug.scsi_debug.q35: PASS (119.02 s)
 (07/12) Host_RHEL.m8.u2.product_av.raw.virtio_blk.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.block_discard_hotplug.local_file: PASS (122.78 s)
 (08/12) Host_RHEL.m8.u2.product_av.raw.virtio_blk.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.block_discard_hotplug.local_file.q35: PASS (130.71 s)
 (09/12) Host_RHEL.m8.u2.product_av.raw.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.block_discard_hotplug.local_file: PASS (132.25 s)
 (10/12) Host_RHEL.m8.u2.product_av.raw.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.block_discard_hotplug.local_file.q35: PASS (127.55 s)
 (11/12) Host_RHEL.m8.u2.product_av.raw.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.block_discard_hotplug.scsi_debug: PASS (127.81 s)
 (12/12) Host_RHEL.m8.u2.product_av.raw.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.block_discard_hotplug.scsi_debug.q35: PASS (118.13 s)

**Comment by qingwangrh** [comment]
@YongxueHong @zhenyzha Please help to review,thanks.

File: qemu/tests/block_discard_hotplug.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        disk_name = storage.get_image_filename(image_params,
+                                               data_dir.get_data_dir())
+
+    timeout = float(params.get("login_timeout", 240))
+    session = vm.wait_for_login(timeout=timeout)
+    plug = BlockDevicesPlug(vm)
+    error_context.context("Hot-plug discarded disk in guest.", logging.info)
+    plug.hotplug_devs_serial(data_tag)
+
+    guest_disk_name = get_guest_discard_disk()
```
Comment (YongxueHong) [inline qemu/tests/block_discard_hotplug.py:95]:
```suggestion
    plug.hotplug_devs_serial(data_tag)
    guest_disk_name = '/dev/' + plug[0]
```
if you call the `plug.hotplug_devs_serial`, you could get the hotplugged disk id by `plug[index]`.

File: qemu/tests/cfg/block_discard_hotplug.cfg [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+      - local_file:
+          force_create_image_stg1 = yes
+          image_size_stg1 = 1G
+          image_name_stg1 = images/stg1
+          blk_extra_params_stg1 += "serial=${disk_serial}"
+          # It need set discard=on the virtio-blk device under pc mode
+          virtio_blk..i440fx:
+            blk_extra_params_stg1 += ",discard=on"
+      - scsi_debug:
+          no virtio_blk
```
Comment (YongxueHong) [inline qemu/tests/cfg/block_discard_hotplug.cfg:29]:
```suggestion
          only virtio_scsi
```
Maybe only virtio_scsi is more matching your case name `scsi_debug`.

File: qemu/tests/block_discard_hotplug.py (hunk 1 above)
Comment (qingwangrh) [reply qemu/tests/block_discard_hotplug.py:95]: Will update

File: qemu/tests/cfg/block_discard_hotplug.cfg (hunk 2 above)
Comment (qingwangrh) [reply qemu/tests/cfg/block_discard_hotplug.cfg:29]: Will update

**Comment by zhenyzha** [comment]
python3 ConfigTest.py --guestname=RHEL.8.3.0 --platform=ppc64le --driveformat=virtio_blk --nicmodel=virtio_net --testcase=block_discard_hotplug --imageformat=raw,qcow2

 (1/6) Host_RHEL.m8.u3.product_av.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads: PASS (1540.15 s)
 (2/6) Host_RHEL.m8.u3.product_av.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.rh_kernel_update: PASS (263.93 s)
 (3/6) Host_RHEL.m8.u3.product_av.raw.virtio_blk.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads: PASS (1479.95 s)
 (4/6) Host_RHEL.m8.u3.product_av.raw.virtio_blk.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.rh_kernel_update: PASS (406.03 s)
 (5/6) Host_RHEL.m8.u3.product_av.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.block_discard_hotplug.local_file: ERROR: Shell command failed: 'mkdir -p /home/test;mkfs.ext4  && mount  /home/test'    (status: 1,    output: 'Usage: mkfs.ext4 [-c|-l filename] [-b block-size] [-C cluster-size]\n\t[-i bytes-per-inode] [-I inode-size] [-J journal-options]\n\t[-G flex-group-size]... (198.03 s)
 (6/6) Host_RHEL.m8.u3.product_av.raw.virtio_blk.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.block_discard_hotplug.local_file: ERROR: Shell command failed: 'mkdir -p /home/test;mkfs.ext4  && mount  /home/test'    (status: 1,    output: 'Usage: mkfs.ext4 [-c|-l filename] [-b block-size] [-C cluster-size]\n\t[-i bytes-per-inode] [-I inode-size] [-J journal-options]\n\t[-G flex-group-size]... (109.77 s)

error msg:
 aexpect.exceptions.ShellCmdError: Shell command failed: 'mkdir -p /home/test;mkfs.ext4  && mount  /home/test'    (status: 1,    output: 'Usage: mkfs.ext4 [-c|-l filename] [-b block-size] [-C cluster-size]\n\t[-i bytes-per-inode] [-I inode-size] [-J journal-options]\n\t[-G flex-group-size] [-N number-of-inodes] [-d root-directory]\n\t[-m reserved-blocks-percentage] [-o creator-os]\n\t[-g blocks-per-group] [-L volume-label] [-M last-mounted-directory]\n\t[-O feature[,...]] [-r fs-revision] [-E extended-option[,...]]\n\t[-t fs-type] [-T usage-type ] [-U UUID] [-e errors_behavior][-z undo_file]\n\t[-jnqvDFSV] device [blocks-count]\n')

**Comment by qingwangrh** [comment]
Patch updated.Please help to review.

File: qemu/tests/cfg/block_discard_hotplug.cfg [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    no Windows
+    type = block_discard_hotplug
+    data_tag = "stg1"
+    images += " stg0 ${data_tag}"
+    force_create_image_stg0 = yes
+    image_size_stg0 = 1G
+    image_name_stg0 = images/stg0
+    drv_extra_params_stg0 += "discard=unmap"
+    boot_drive_stg1 = no
+    disk_serial = TARGET_DISK1
```
Comment (YongxueHong) [inline qemu/tests/cfg/block_discard_hotplug.cfg:11]: Looks like we don't need `disk_serial = TARGET_DISK1` any more.

File: qemu/tests/block_discard_hotplug.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    def get_scsi_debug_disk(guest_session=None):
+        """"
+        Get scsi debug disk on host or on guest which created as scsi-block.
+        """
+        cmd = "lsblk -S -n -p|grep scsi_debug"
+
+        if guest_session:
+            status, output = guest_session.cmd_status_output(cmd)
+        else:
+            status, output = process.getstatusoutput(cmd)
```
Comment (YongxueHong) [inline qemu/tests/block_discard_hotplug.py:44]:
```suggestion
        status, output = process.getstatusoutput(cmd)
```
Better to remove the part of guest session then modify the docstring `Get scsi debug disk on host which created as scsi-block.`

**Comment by zhenyzha** [comment]
(1/2) Host_RHEL.m8.u3.product_av.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.block_discard_hotplug.
local_file: PASS (281.49 s)
 (2/2) Host_RHEL.m8.u3.product_av.raw.virtio_blk.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.block_discard_hotplug.lo
cal_file: PASS (245.45 s)
 (1/2) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.block_discard_hotplug.local_file: PASS (307.97 s)
 (2/2) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.block_discard_hotplug.scsi_debug: PASS (257.94 s)

The current inspection method will not trigger the above problem:error msg:
`aexpect.exceptions.ShellCmdError: Shell command failed...`
I will explain the specific reason for the above problem in the bug.

**Review by YongxueHong — APPROVED** [review]

File: qemu/tests/block_discard_hotplug.py (hunk 4 above)
Comment (qingwangrh) [reply qemu/tests/block_discard_hotplug.py]: updated

File: qemu/tests/cfg/block_discard_hotplug.cfg (hunk 3 above)
Comment (qingwangrh) [reply qemu/tests/cfg/block_discard_hotplug.cfg]: updated

**Comment by qingwangrh** [comment]
> (1/2) Host_RHEL.m8.u3.product_av.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.block_discard_hotplug.
> local_file: PASS (281.49 s)
> (2/2) Host_RHEL.m8.u3.product_av.raw.virtio_blk.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.block_discard_hotplug.lo
> cal_file: PASS (245.45 s)
> (1/2) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.block_discard_hotplug.local_file: PASS (307.97 s)
> (2/2) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.block_discard_hotplug.scsi_debug: PASS (257.94 s)
> 
> The current inspection method will not trigger the above problem:error msg:
> `aexpect.exceptions.ShellCmdError: Shell command failed...`
> I will explain the specific reason for the above problem in t

**Comment by zhenyzha** [comment]
python3 ConfigTest.py --testcase=block_discard_hotplug --guestname=RHEL.8.3.0 --driveformat=virtio_blk,virtio_scsi --imageformat=raw,qcow2  

```
(01/14) Host_RHEL.m8.u3.product_av.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.unattended_install.c
drom.extra_cdrom_ks.default_install.aio_threads: PASS (1725.94 s)
 (02/14) Host_RHEL.m8.u3.product_av.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.rh_kernel_update: PASS (277.48 s)
 (03/14) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads: PASS (1694.69 s)
 (04/14) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.rh_kernel_update: PASS (336.78 s)
 (05/14) Host_RHEL.m8.u3.product_av.raw.virtio_blk.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads: PASS (1645.27 s)
 (06/14) Host_RHEL.m8.u3.product_av.raw.virtio_blk.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.rh_kernel_update: PASS (413.20 s)
 (07/14) Host_RHEL.m8.u3.product_av.raw.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads: PASS (1672.61 s)
 (08/14) Host_RHEL.m8.u3.product_av.raw.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.rh_kernel_update: PASS (576.58 s)
 (09/14) Host_RHEL.m8.u3.product_av.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.block_discard_hotplug.local_file: PASS (326.43 s)
 (10/14) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.block_discard_hotplug.local_file: PASS (299.15 s)
 (11/14) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.block_discard_hotplug.scsi_debug: PASS (286.92 s)
 (12/14) Host_RHEL.m8.u3.product_av.raw.virtio_blk.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.block_discard_hotplug.local_file: PASS (281.20 s)
 (13/14) Host_RHEL.m8.u3.product_av.raw.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.block_discard_hotplug.local_file: PASS (290.99 s)
 (14/14) Host_RHEL.m8.u3.product_av.raw.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.block_discard_hotplug.scsi_debug: PASS (240.23 s)
```

LGTM.Acked-by: Zhenyu Zhang <zhenyzha@redhat.com>

**Review by zhenyzha — APPROVED** [review]

**Pull request merged**